core/modules/user/user.module

```diff
@@ -1755,6 +1800,17 @@ function user_login_final_validate($form, &$form_state) {
       }
     }
     else {
+      switch (config_get('user.settings', 'user_login_method')) {
+        case USER_LOGIN_USERNAME_ONLY:
+          form_set_error('name', t('Sorry, unrecognized username or password. <a href="@password">Have you forgotten your password?</a>', array('@password' => url('user/password'))));
+          break;
+        case USER_LOGIN_EMAIL_ONLY:
+          form_set_error('name', t('Sorry, unrecognized e-mail address or password. <a href="@password">Have you forgotten your password?</a>', array('@password' => url('user/password'))));
+          break;
+        case USER_LOGIN_USERNAME_OR_EMAIL:
+          form_set_error('name', t('Sorry, unrecognized username, e-mail address, or password. <a href="@password">Have you forgotten your password?</a>', array('@password' => url('user/password'))));
+        break;
+      }
       form_set_error('name', t('Sorry, unrecognized username or password. <a href="@password">Have you forgotten your password?</a>', array('@password' => url('user/password', array('query' => array('name' => $form_state['values']['name']))))));
```

Comment:
Looks like this line should be deleted, it will throw a second error on the same form element.